[18.0][MIG] web_dark_mode: Migration to 18.0

ljmnoonan opened the pull request:
This PR includes and supersedes #3267 with the permission of @len-foss.

Migrates `web_dark_mode` from 16 to 18 and overhauls the SCSS not only to look better, but also to correspond to Odoo's native structure as seen in the `web` module.

As with the previous version, its scope is limited to the backend, and so has no effect on POS or website.

Also, I have submitted #3322 to make `web_theme_classic` get along with this dark mode hope to do the same for `web_responsive` soon.

Looking forward to any feedback!

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/3019d6c3-8e37-4c70-ae1c-0e727231267f" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/28f04d89-6053-4140-b171-ca749dd6bcf6" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/ad1b94a8-4d50-448a-9b05-e7f67ef17e41" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/a7cca4ef-fb9f-4ffc-a4a1-87ef1f5fd610" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/09a3bab8-dfc4-4872-9dd8-2ab6e5a789b4" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/0fb26e52-c5a9-4b8b-a727-beedd099d9d7" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/1e40a1f5-1b50-4efb-817c-4b1770990b57" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/df3d2c6c-c449-4d46-8127-ae84e8f4cb14" />


<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/4f110586-268a-4682-ba13-5d344aecf37d" />


<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/4284ebae-4b64-40b7-9f1c-9a01a53bc5da" />

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/ac3cc713-18a9-4fe3-a9ac-82b54021c135" />

fkantelberg committed:
[ADD][16.0] Module web_dark_mode

Commit:
[UPD] Update web_dark_mode.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

GoodERPJeff committed:
Added translation using Weblate (Chinese (zh))

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (4 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/es_AR/

fkantelberg committed:
[FIX] web_dark_mode: Fix uninstallation

OCA-git-bot committed:
web_dark_mode 16.0.1.0.1

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (4 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/hr/

tate11 committed:
Added translation using Weblate (Turkish)

tate11 committed:
Translated using Weblate (Turkish)

Currently translated at 100.0% (4 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/tr/

Ivorra78 committed:
Added translation using Weblate (Spanish)

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (4 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/es/

OCA-git-bot committed:
[UPD] README.rst

gregory-moka committed:
Added translation using Weblate (French)

gregory-moka committed:
Translated using Weblate (French)

Currently translated at 50.0% (2 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/fr/

fkantelberg committed:
[FIX] web_dark_mode: Add fields to field lists

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: web-16.0/web-16.0-web_dark_mode
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_dark_mode/it/

len-foss committed:
[MIG] web_dark_mode: Migration to 18.0

len-foss committed:
[IMP] web_dark_mode: pre-commit execution

ljmnoonan committed:
[IMP] web_dark_mode Set dark theme to immitate VSCode Modern Dark
Also, change asset bundle loading from 'prepend' to 'before' This is in line with Odoo's own way of doing it.

ljmnoonan committed:
[IMP] web_dark_mode Change navbar from purple to dark
Also get rid of the bottom border.

ljmnoonan committed:
[IMP] web_dark_mode Fix tint and shade color functions

ljmnoonan committed:
[IMP] web_dark_mode Add lazy loading asset bundle
If you go to any activity view and hover over an empty cell you will notice that
without this fix the the on hover background of empty cells (.o_activity_empty_cell) set by
$component-active-bg which is itself set as mix($o-action, $o-gray-100, 20%) through $o-component-active-bg
completely ignores the overrides we set in primary_variables.dark.scss.
This is because the assets for that view are lazy loaded, so we need to include our override
bundles in the lazy bundle.

ljmnoonan committed:
[IMP] web_dark_mode Override BS variables
Just general customizations of BS variables to bring them in line with our style

ljmnoonan committed:
[IMP] web_dark_mode Improve notebook tab styling

ljmnoonan committed:
[IMP] web_dark_mode Adjust tag and colorlist colors

ljmnoonan committed:
[IMP] web_dark_mode Adjust calendar coloring
Add secondary_variables override to get rid of colors that don't jive well with the dark background.
Unfortunately this does introduce an issue where if you have many calendar users, you can run out of colors
This issue exists in the enterprise dark mode as well, so hopefully it will be fixed.

ljmnoonan committed:
[IMP] web_dark_mode Adjust the styles of custom properties

ljmnoonan committed:
[IMP] web_dark_mode Tweak the search bar
Set the background color to be darker and tone down the border around it.
The styling for the border is borrwed from the style of .o-mail-Composer-bg

ljmnoonan committed:
[IMP] web_dark_mode Set image background to white

ljmnoonan committed:
[IMP] web_dark_mode Tweak kanban controller
Set attachment image background to white and fix the column background when hovering with a draggable card

ljmnoonan committed:
[IMP] web_dark_mode: Add myself to CONTRIBUTORS.md
I don't want to but my email on a public repo, so I just linked my github page instead.
I could not find any guideline against this, but please correct me if there is.

mathben commented:
Thanks, it works for me.

Did you test the dashboard? The background is white.

ljmnoonan commented:
> Did you test the dashboard? The background is white.

The dashboard is, unfortunately, mostly canvas elements that I presume are rendered by JavaScript somewhere. I didn't look into it too much this time, but hopefully I'll be able to get it sorted later.
This problem existed in version 16 as well.

len-foss commented:
FWIW, I have this in `lambdao_web`'s `ui.scss`:
```css
.app-menu-container {
  background-image: url(/lambdao_web/static/src/img/logo_grey.svg) !important;
  background-position: bottom right !important;
  background-repeat: no-repeat !important;
  background-size: 10% !important;
  background-color: #899499 !important;
}
```
and this in `ui_dark.scss`:
```css
 .app-menu-container {
   background-color: #213330 !important;
 }
```
in any case, you don't need any JS to fix that. I had put a bunch of theme fixes for all the things I found egregious, this one being the worst offender :-)

ljmnoonan commented:
@len-foss 
Thank you for pointing that out! I'll try to get it fixed.

ljmnoonan commented:
@len-foss, `app-menu-container` class is from `web_responsive`, the app switcher dashboard as far as I can tell. I am working on a PR for a dark mode for web responsive too, but @mathben and I were referring to the dashboards module earlier.

<img width="1706" height="1440" alt="image" src="https://github.com/user-attachments/assets/b1f46b3c-bd98-4a0f-ae36-4cb268f8367c" />

Now, I am not very good with javascript, but it does seem that this is all being rendered in [o_spreadsheet.js](https://github.com/odoo/odoo/blob/18.0/addons/spreadsheet/static/src/o_spreadsheet/o_spreadsheet.js) and the styles that are controlled by SCSS are deliberately set to stay in light mode even when dark mode is active in [o_spreadsheet_extended.dark.scss](https://github.com/odoo/odoo/blob/18.0/addons/spreadsheet/static/src/o_spreadsheet/o_spreadsheet_extended.dark.scss) the first lines of which contain a comment mentioning that odoo does not currently have an adequate toolchain to support darkmode in spreadsheet.
I am not going to pursue this any further at the moment as it looks like a ton of work, and something that should actually be contributed upstream to odoo anyways.

len-foss commented:
Oh sorry for the confusion. After the previous message I saw that I had a bunch of "fixes" to the dark mode, but without proper doc on chat it was. I agree with your analysis :+1:

ljmnoonan committed:
[IMP] web_dark_mode Change thead active (column filter) bg for lists in forms
It was being set to $o-form-lightsecondary which I set lighter, so this tones it down a bit.

mathben commented:
Can you review please @legalsylvain @yajo @LoisRForgeFlow 
Thanks you!

ljmnoonan commented:
@len-foss, would you mind reviewing this so we could maybe get some movement? 
Thank you!

len-foss reviewed (approved):
LGTM
In any case, I don't see any good reason to stall the PR, except demoralization of potential contributors. 

For completeness, there are some issues left (I guess left from the implementation of dark mode in CE that is severely lacking) -- checking on the runboat, I could see the following, by example.

<img width="1638" height="811" alt="Screenshot_20251230_182642" src="https://github.com/user-attachments/assets/dd411363-1328-4a0e-847e-3874dc567600" />

pedrobaeza commented:
> For completeness, there are some issues left (I guess left from the implementation of dark mode in CE that is severely lacking) -- checking on the runboat, I could see the following, by example.

I think that's because `web_classic_theme`. Please uninstall it. I see that module as very harmless, as it's very bad for the visual health and ergonomics.

ljmnoonan commented:
@len-foss and @pedrobaeza 
I fixed the styles causing that eyesore in `web_theme_classic` in #3322. Would you like me to rebase so this becomes visible in runboat?
Thanks for reviewing!

ljmnoonan commented:
> I see that module as very harmless, as it's very bad for the visual health and ergonomics.

Also, btw, I have #3394 open to make `web_theme_classic` a user or even device option, so it isn't do all or nothing

pedrobaeza reviewed (approved):
No need, let's move with this:

/ocabot migration web_dark_mode

/ocabot merge nobump

@len-foss the problem about merging without proper reviewing from trusted reviewers is to lower the quality perception of people using OCA modules. I try to push some of them when others that are more or less familiar do reviews, but there's also a trend of some people "spitting" migration PRs (or even imp) not being even properly tested in their end.

pedrobaeza commented:
@ljmnoonan but are you using `web_classic_theme` in your deployments? As said, I think keeping such purple background and hard borders is very harmless for the view health. I can agree default theme may be too "light" in some parts, but we can work on that not on that extreme way. Even that changes can be included in `web_responsive`, and to have several themes on it (the system is already prepared for that).

len-foss commented:
@ljmnoonan oh, sorry for the noise again -- I just hopped on runboat without much more checks (and I was indeed confused as to how such an obvious issue could have been overlooked). Thanks for your work!

ljmnoonan commented:
@pedrobaeza, I don't have any real production deployments yet as I am just learning in my free time with the goal of eventually setting this up at the place I work (a lumber yard).  I do like having `web_theme_classic` available though, but I also like to be able to turn it off. Really, it is most helpful for custom properties in tasks and such. I did a lot of cleanup on it in my earlier PR too, so it now it overrides the relevant css at the same level it was defined at, rather than the old approach which "chased" specific fields in specific contexts, resulting in inconsistencies and conflicts. 
<details>
<summary> A screenshot with the new style </summary>
<img width="1920" height="1032" alt="image" src="https://github.com/user-attachments/assets/7d82687c-aa7a-4a31-8924-bb2a4c2333bf" />
</details>

As for the idea of moving `web_dark_mode` to `web_responsive`, I totally agree that it would make a lot of sense.  However, as things are right now in `web_responsive` the "themes", which only actually affect the app switcher, are implemented in a rather convoluted way using both scss and css variables switching based on an HTML data attribute rather than odoo's native backend asset bundling logic specifically built for dark mode and for which many core odoo apps have dedicated asset bundles. This is something I would like to discuss and work on, so I opened #3387 , but have so far got no input.

Thank you for the approval on this PR too!

pedrobaeza commented:
@ljmnoonan that's at least less aggressive, but I suppose that only in dark mode. What I mean is to touch the light part in `web_responsive` for not requiring `web_classic_theme` at all. Some things should be sacrificed, but finding a good compromise between ergonomics and user awareness.

And congratulations for your work.

ljmnoonan commented:
@pedrobaeza Ah, I see what you mean; integrate dome of the functionality of `web_theme_classic` into `web_responsive`. Maybe just bottom borders? Maybe revert to default in list view too.
<img width="1920" height="1032" alt="image" src="https://github.com/user-attachments/assets/b789cd47-3b6a-4217-a579-06665cf922eb" />
